src/Components/Components/src/NavigationManager.cs

```diff
@@ -102,7 +103,8 @@ public void NavigateTo(string uri, bool forceLoad = false)
         /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI
         /// (as returned by <see cref="BaseUri"/>).</param>
         /// <param name="forceLoad">If true, bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.</param>
-        protected abstract void NavigateToCore(string uri, bool forceLoad);
+        /// <param name="replace">If true, will replace the uri in the current browser history state, instead of pushing the new uri onto the browser history stack</param>
+        protected abstract void NavigateToCore(string uri, bool forceLoad, bool replace);
```

Comment:
Thanks for the contribution, @MariovanZeist! This will be a great feature to have.

We will need to find a way to make this change without it being a breaking change. The way it's implemented here would mean that all existing subclasses of `NavigationManager` would become broken because they don't take a `replace` parameter.

One way this could be done would be like this:

```cs
// Leave this alone for back-compatibility
protected abstract void NavigateToCore(string uri, bool forceLoad);

protected virtual void NavigateToCore(string uri, bool forceLoad, bool replace)
{
    // Default implementation ignores "replace"
    NavigateToCore(uri, forceLoad);
}
```

Then, subclasses of `NavigationManager` can optionally override the 3-parameter overload of `NavigateToCore` and provide a more specialised implementation that respects `replace`. Other subclasses of `NavigationManager` will continue to work, but will just ignore the third parameter.

Another option would be that the 3-parameter overload of `NavigateToCore` should throw a `NotSupportedException` in the abstract base class rather than ignoring the third parameter. In this case, we'd want to have two separate overloads of `NavigateTo` - the existing one unchanged for back-compatibility, and a new three-parameter one that calls the three-parameter overload of `NavigateToCore`.

Finally, another design point to consider is making this more extensible in the future. What if we need a 4th parameter one day? One way to address this is defining some `struct NavigationOptions` that contains the `bool Replace` flag, and have that as the new third parameter. Then in the future we could add more flags or sub-parameters into that struct.

Does this make sense?

Reply:
Thanks @SteveSandersonMS for considering my PR, 
Yes I totally agree with your proposed changes, I wasn't considering binary compatibility but I should have when I created this PR.

I personally prefer the `struct NavigationOptions` version, just because of the extensibility. 

so my suggestion would something like :

```cs
//Add options class
public class NavigationOptions
{
    public bool ForceLoad { get; set;}
    public bool Replace { get; set; }
}

//Changes to NavigationManager

//keep this for backward compatibility
 public void NavigateTo(string uri, bool forceLoad = false)
{
  AssertInitialized();
  NavigateToCore(uri, new NavigationOptions { ForceLoad = forceLoad} );
}

//Add overload
public void NavigateTo(string uri, NavigationOptions options)
{
  AssertInitialized();
  NavigateToCore(uri, options);
}


//keep this for backward compatibility
protected abstract void NavigateToCore(string uri, bool forceLoad);

protected virtual void NavigateToCore(string uri, NavigationOptions options)
{
    // Default implementation ignores "replace"
    NavigateToCore(uri, options.ForceLoad);
}

```
Would these changes be acceptable?.

And does this break binary compatibility later if we decide to extend NavigationOptions ?